### PR TITLE
[reverted] python2 and python3: reproducible builds

### LINKED ISSUE #107951: python3 library build is not deterministic
**Describe the bug**

When building python3 twice, the result is not always bit-by-bit the same. This makes it harder to audit whether binaries/images have been built correctly. For more information on why this is useful see https://reproducible-builds.org .

**To Reproduce**

`nix-build '<nixpkgs>' -A python3 --check` 

**Additional context**

Notably, the ordering of some symbols (`​_Py_Mangle`, `​PyEval_GetGlobals`, several `_PyErr` ones) is not stable.

Setting `enableParallelBuilding` to false doesn't help.

### PR DESCRIPTION
**Pull request by FRidh**
Achieve reproducible builds of the interpreter. Note this meant
disabling optimizations again that were introduced in https://github.com/NixOS/nixpkgs/pull/84072.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

todo:
- [x] changelog
- [x] fix comments

### CODE REVIEW

**Comment by FRidh** [comment]
After testing I did some reordering resulting in a mass rebuild. I have not tested this new commit (its ongoing).

**Comment by FRidh** [comment]
Non-optimized bytecode is now removed, which means `__debug__ = false`.

I don't know which .pyc is actually picked up when bytecode is available for multiple optimization levels.

File: pkgs/development/interpreters/python/cpython/2.7/default.nix [hunk 1]
```diff
@@ -260,14 +277,18 @@ in with passthru; stdenv.mkDerivation ({
...
-      ''
-        # Determinism: rebuild all bytecode
+      '' + optionalString stripBytecode ''
+        # Determinism: deterministic bytecode
+        # First we delete all old bytecode.
+        find $out -name "*.pyc" -delete
+        '' + optionalString rebuildBytecode ''
+        # Then, we build for the two optimization levels.
+        # We do not build unoptimized bytecode, because its not entirely deterministic yet.
+        # Python 3.7 implements PEP 552, introducing support for deterministic bytecode.
```
Comment (FRidh) [inline pkgs/development/interpreters/python/cpython/2.7/default.nix:287]: Irrelevant here, python 2.

File: pkgs/development/interpreters/python/cpython/2.7/default.nix [hunk 2]
```diff
@@ -260,14 +277,18 @@ in with passthru; stdenv.mkDerivation ({
...
+        # compileall by used this checked-hash method by default when `SOURCE_DATE_EPOCH` is set.
         # We exclude lib2to3 because that's Python 2 code which fails
-        # We rebuild three times, once for each optimization level
-        find $out -name "*.py" | $out/bin/python -m compileall -q -f -x "lib2to3" -i -
-        find $out -name "*.py" | $out/bin/python -O -m compileall -q -f -x "lib2to3" -i -
-        find $out -name "*.py" | $out/bin/python -OO -m compileall -q -f -x "lib2to3" -i -
+        find $out -name "*.py" | ${pythonForBuildInterpreter} -O  -m compileall -q -f -x "lib2to3" -i -
```
Comment (FRidh) [inline pkgs/development/interpreters/python/cpython/2.7/default.nix:290]: python 2 stores optimized bytecode in .pyo and stores only one level, not two. Thus we need to choose.

File: pkgs/development/interpreters/python/cpython/2.7/default.nix [hunk 3]
```diff
@@ -27,7 +27,10 @@
 , sha256
 , passthruFun
 , static ? false
-, enableOptimizations ? (!stdenv.isDarwin)
+, stripBytecode ? reproducible
+, rebuildBytecode ? true #stdenv.hostPlatform == stdenv.buildPlatform
```
Comment (FRidh) [inline pkgs/development/interpreters/python/cpython/2.7/default.nix:31]: should check cross

File: pkgs/development/interpreters/python/cpython/default.nix [hunk 4]
```diff
@@ -291,18 +299,19 @@ in with passthru; stdenv.mkDerivation {
...
+    # Determinism: deterministic bytecode
+    # First we delete all old bytecode.
+    find $out -type d -name __pycache__ -print0 | xargs -0 -I {} rm -rf "{}"
+    '' + optionalString rebuildBytecode ''
+    # Then, we build for the two optimization levels.
+    # We do not build unoptimized bytecode, because its not entirely deterministic yet.
     # Python 3.7 implements PEP 552, introducing support for deterministic bytecode.
-    # This is automatically used when `SOURCE_DATE_EPOCH` is set.
-    find $out -name "*.py" | ${pythonForBuildInterpreter}     -m compileall -q -f -x "lib2to3" -i -
+    # compileall by used this checked-hash method by default when `SOURCE_DATE_EPOCH` is set.
```
Comment (FRidh) [inline pkgs/development/interpreters/python/cpython/default.nix:311]: something went wrong with this sentence

File: pkgs/development/interpreters/python/cpython/2.7/default.nix [hunk 5]
```diff
@@ -36,12 +39,26 @@ assert x11Support -> tcl != null
                   && xlibsWrapper != null
                   && libX11 != null;
 
+assert stdenv.lib.assertMsg (enableOptimizations -> (!stdenv.isDarwin))
```
Comment (FRidh) [inline pkgs/development/interpreters/python/cpython/2.7/default.nix:42]: its actually clang that is the issue https://github.com/NixOS/nixpkgs/pull/103715

**Review by raboof — COMMENTED** [review]
I'm thrilled this indeed seems to make python reproducible. On the other hand, I can see how disabling the optimizations is painful (not sure about the stripping)...

**Comment by drewrisinger** [comment]
I looked through this PR briefly, and I was wondering if it's possible to have the optimizations as a deterministic input. That would probably be a separate PR, though.

Something like:
1. Generate optimizations as part of a (new) derivation.
2. Use optimizations as input to the python interpreter build. This could either be an optional or default argument, I'm open to either.
3. Output a deterministic, but optimized, python interpreter.

Honestly, I haven't looked enough into the python optimization process to know if this is possible, it just seemed like a best-of-both-worlds scenario.

**Comment by tomberek** [comment]
Result of `nixpkgs-review pr 107965` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>python3</li>
  </ul>
</details>

**Comment by FRidh** [comment]
I will go ahead with this soon, at least before 21.05. Need to make some minor changes, and add release notes and docs.

**Comment by FRidh** [comment]
- comment needs to be fixed
- documentation needs to be updated
- release notes should be added

**Pull request merged**